grudge/models/euler.py

```diff
@@ -0,0 +1,364 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+
+from abc import ABCMeta, abstractmethod
+
+from dataclasses import dataclass
+from arraycontext import (
+    thaw,
+    dataclass_array_container,
+    with_container_arithmetic
+)
+
+from meshmode.dof_array import DOFArray
+
+from grudge.discretization import DiscretizationCollection
+from grudge.dof_desc import DOFDesc, DISCR_TAG_BASE, as_dofdesc
+from grudge.models import HyperbolicOperator
+from grudge.trace_pair import TracePair
+
+from pytools.obj_array import make_obj_array
+
+import grudge.op as op
+
+
+# {{{ Array container for the Euler model
+
+@with_container_arithmetic(bcast_obj_array=False,
+                           bcast_container_types=(DOFArray, np.ndarray),
+                           matmul=True,
+                           rel_comparison=True)
+@dataclass_array_container
+@dataclass(frozen=True)
+class EulerField:
+    mass: DOFArray
+    energy: DOFArray
+    momentum: np.ndarray
+
+    @property
+    def array_context(self):
+        return self.mass.array_context
+
+    @property
+    def dim(self):
+        return len(self.momentum)
+
+# }}}
+
+
+# {{{ Predefined initial conditions for the Euler model
+
+def vortex_initial_condition(x_vec, t=0):
+    """Initial condition adapted from Section 2 (equation 2) of:
+
+    - K. Mattsson, M. Sv\"{a}rd, M. Carpenter, and J. Nordstr\"{o}m (2006).
+    High-order accurate computations for unsteady aerodynamics.
+    [DOI](https://doi.org/10.1016/j.compfluid.2006.02.004).
+    """
+    mach = 0.5    # Mach number
+    _x0 = 5
+    epsilon = 1   # vortex strength
+    gamma = 1.4
+    x, y = x_vec
+    actx = x.array_context
+
+    fxyt = 1 - (((x - _x0) - t)**2 + y**2)
+    expterm = actx.np.exp(fxyt/2)
+
+    u = 1 - (epsilon*y/(2*np.pi)) * expterm
+    v = ((epsilon*(x - _x0) - t)/(2*np.pi)) * expterm
+
+    velocity = make_obj_array([u, v])
+    mass = (
+        1 - ((epsilon**2 * (gamma - 1) * mach**2)/(8*np.pi**2)) * actx.np.exp(fxyt)
+    ) ** (1 / (gamma - 1))
+    p = (mass ** gamma)/(gamma * mach**2)
+
+    return primitive_to_conservative_vars((mass, velocity, p), gamma=gamma)
+
+# }}}
+
+
+# {{{ Variable transformation and helper routines
+
+def conservative_to_primitive_vars(cv_state, gamma=1.4):
+    """Converts from conserved variables (density, momentum, total energy)
+    into primitive variables (density, velocity, pressure).
+
+    :arg cv_state: A :class:`EulerField` containing the conserved
+        variables.
+    :arg gamma: The isentropic expansion factor for a single-species gas
+        (default set to 1.4).
+    :returns: A :class:`Tuple` containing the primitive variables:
+        (density, velocity, pressure).
+    """
+    rho = cv_state.mass
+    rho_e = cv_state.energy
+    rho_u = cv_state.momentum
+    u = rho_u / rho
+    p = (gamma - 1) * (rho_e - 0.5 * sum(rho_u * u))
+
+    return rho, u, p
```

Comment:
Have a dataclass for these, too?

Reply:
Ehh I think that might be a bit excessive for these. They're only really used internally (except if explicitly requested via the documented helper routine `conservative_to_primitive_vars` for plotting purposes).

Reply:
Fair enough. Also easy to hack in later.